crates/pallet-domains/src/block_tree.rs

```diff
@@ -93,24 +109,34 @@ pub(crate) fn execution_receipt_type<T: Config>(
         Ordering::Greater => ReceiptType::Rejected(RejectedReceiptType::InFuture),
         Ordering::Equal => ReceiptType::Accepted(AcceptedReceiptType::NewHead),
         Ordering::Less => {
+            // Reject receipt that already pruned/confirmed
             let oldest_receipt_number =
                 head_receipt_number.saturating_sub(T::BlockTreePruningDepth::get());
-            let already_exist = BlockTree::<T>::get(domain_id, receipt_number)
-                .contains(&execution_receipt.hash::<DomainHashingFor<T>>());
-
             if receipt_number < oldest_receipt_number {
-                // Receipt already pruned
-                ReceiptType::Rejected(RejectedReceiptType::Pruned)
-            } else if !already_exist {
-                // Create new branch
-                ReceiptType::Accepted(AcceptedReceiptType::NewBranch)
-            } else if receipt_number == head_receipt_number {
-                // Add comfirm to the current head receipt
-                ReceiptType::Accepted(AcceptedReceiptType::CurrentHead)
-            } else {
-                // Add comfirm to a non-head receipt
-                ReceiptType::Stale
+                return ReceiptType::Rejected(RejectedReceiptType::Pruned);
+            }
+
+            // Reject receipt that try to create new branch in the block tree
+            let already_exist = is_receipt_exist::<T>(
+                domain_id,
+                receipt_number,
+                execution_receipt.hash::<DomainHashingFor<T>>(),
+            );
+            if !already_exist {
+                return ReceiptType::Rejected(RejectedReceiptType::NewBranch);
             }
+
+            // Add confirm to the head receipt that added in the current block or it is
+            // the genesis receipt
+            let head_receipt_extended = HeadReceiptExtended::<T>::get(domain_id);
```

Comment:
we do not need this Storage.
There is `SuccessfulBundles` storage. For the first bundle with ER, should be `head_receipt_number + 1` and not a first bundle, then it should point to `head_receipt_number` since `head_receipt_number` is updated when the first bundle is accepted

Reply:
I have considered this, but would prefer to have an explicit `HeadReceiptExtended` to give more context here

Reply:
This is inconsistant then. Some places we use `SuccessfulBundles` and other places we use this storage to do the same.
Given right amount of context through comments is good IMO rather than introducing new Storage that basically do the same thing

Reply:
`SuccessfulBundles` and `HeadReceiptExtended` are not really used to do the same thing. Yeah, we did use `SuccessfulBundles::<T>::get(domain_id).is_empty()` in the `submit_bundle` call to determine if it is the first bundle and we can also use it here, but that will bring a lot of implication about relation between bundle and ER, that   required more comment to explain and more brain power of the reader to understand.

Thus, I would prefer to use `HeadReceiptExtended`, which is explicit, simple, easy to understand, and give enough context here. And at the end, `HeadReceiptExtended` will be removed if we going to have multiple ER in one bundle.

Reply:
Then lets be consistent and use this storage to determine the first bundle then

Reply:
`SuccessfulBundles::<T>::get(domain_id).is_empty()` is more straightforward to determine the first bundle, they are different thing we don't need to keep the consistency. And again the notion of one ER with one bundle will be removed, along with the `HeadReceiptExtended`, so I don't want to use it in more other places.

Reply:
How is this `if !HeadReceiptExtended::<T>::get(domain_id){}` different from `SuccessfulBundles::<T>::get(domain_id).is_empty()` when determining the first bundle when these two are related to ER.

I dont want to extend this conversation further on this topic. Lets move one 🤷🏼

Reply:
Because one is for bundle and another is for ER, and their relation will be changed soon